backend/src/schema/resolvers/users.spec.js

```diff
@@ -11,50 +11,39 @@ afterEach(async () => {
 })
 
 describe('users', () => {
-  describe('CreateUser', () => {
-    const mutation = `
-      mutation($name: String, $password: String!, $email: String!) {
-        CreateUser(name: $name, password: $password, email: $email) {
-          id
-        }
-      }
-    `
-    describe('given valid password and email', () => {
-      const variables = {
-        name: 'John Doe',
-        password: '123',
-        email: '123@123.de',
-      }
+  describe('User', () => {
+    describe('query by email address', () => {
+      beforeEach(async () => {
+        await factory.create('User', { name: 'Johnny', email: 'any-email-address@example.org' })
+      })
 
-      describe('unauthenticated', () => {
-        beforeEach(async () => {
-          client = new GraphQLClient(host)
-        })
+      const query = `query($email: String) { User(email: $email) { name } }`
+      const variables = { email: 'any-email-address@example.org' }
+      beforeEach(() => {
+        client = new GraphQLClient(host)
+      })
 
-        it('is not allowed to create users', async () => {
-          await expect(client.request(mutation, variables)).rejects.toThrow('Not Authorised')
-        })
+      it('is forbidden', async () => {
+        await expect(client.request(query, variables)).rejects.toThrow('Not Authorised')
       })
 
-      describe('authenticated admin', () => {
+      describe('as admin', () => {
         beforeEach(async () => {
-          const adminParams = {
+          const userParams = {
             role: 'admin',
             email: 'admin@example.org',
             password: '1234',
           }
-          await factory.create('User', adminParams)
-          const headers = await login(adminParams)
+          const factory = Factory()
+          await factory.create('User', userParams)
+          const headers = await login(userParams)
           client = new GraphQLClient(host, { headers })
         })
 
-        it('is allowed to create new users', async () => {
-          const expected = {
-            CreateUser: {
-              id: expect.any(String),
-            },
-          }
-          await expect(client.request(mutation, variables)).resolves.toEqual(expected)
+        it('is permitted', async () => {
+          await expect(client.request(query, variables)).resolves.toEqual({
+            User: [{ name: 'Johnny' }],
+          })
         })
       })
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-03T18:37:58Z" title="Wednesday, July 3rd 2019, 8:37:58 pm +02:00">Jul 3, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -11,51 +11,32 @@ afterEach(async () => {
 })
 
 describe('users', () => {
-  describe('CreateUser', () => {
-    const mutation = `
-      mutation($name: String, $password: String!, $email: String!) {
-        CreateUser(name: $name, password: $password, email: $email) {
-          id
-        }
-      }
-    `
-    describe('given valid password and email', () => {
-      const variables = {
-        name: 'John Doe',
-        password: '123',
-        email: '123@123.de',
-      }
+  describe('User', () => {
+    const query = `query($email: String) { User(email: $email) { id } }`
+    const variables = { email: 'any-email-address@example.org' }
+    beforeEach(() => {
+      client = new GraphQLClient(host)
+    })
 
-      describe('unauthenticated', () => {
-        beforeEach(async () => {
-          client = new GraphQLClient(host)
-        })
+    it('is forbidden', async () => {
+      await expect(client.request(query, variables)).rejects.toThrow('Not Authorised')
+    })
 
-        it('is not allowed to create users', async () => {
-          await expect(client.request(mutation, variables)).rejects.toThrow('Not Authorised')
-        })
+    describe('as admin', () => {
+      beforeEach(async () => {
+        const userParams = {
+          role: 'admin',
+          email: 'admin@example.org',
+          password: '1234',
+        }
+        const factory = Factory()
+        await factory.create('User', userParams)
+        const headers = await login(userParams)
+        client = new GraphQLClient(host, { headers })
       })
 
-      describe('authenticated admin', () => {
-        beforeEach(async () => {
-          const adminParams = {
-            role: 'admin',
-            email: 'admin@example.org',
-            password: '1234',
-          }
-          await factory.create('User', adminParams)
-          const headers = await login(adminParams)
-          client = new GraphQLClient(host, { headers })
-        })
-
-        it('is allowed to create new users', async () => {
-          const expected = {
-            CreateUser: {
-              id: expect.any(String),
-            },
-          }
-          await expect(client.request(mutation, variables)).resolves.toEqual(expected)
-        })
+      it('is permitted', async () => {
+        await expect(client.request(query, variables)).resolves.toEqual({ User: [] })
```


would it be better to actually create a user for these test cases, or you are happy that it either throws an authorization error, or not?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-07-03T20:09:55Z" title="Wednesday, July 3rd 2019, 10:09:55 pm +02:00">Jul 3, 2019</time>_

---


yes, correct. It's a refactoring but I think I can easily add more meaning to the test.. hang on